File: Algorithm/401. Binary Watch.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-09-20 00:11:26
# @Last modified by:   WuLC
# @Last Modified time: 2016-09-20 00:12:27
# @Email: liangchaowu5@gmail.com

# backtracking, bit manipulation
class Solution(object):
    def readBinaryWatch(self, num):
        """
        :type num: int
        :rtype: List[str]
        """
        if num > 8:
            return []
        result = []
        for h in xrange(num+1):
            if h == 4: break
            if num - h > 6: continue
            hours, minutes = [], []
            self.get_hour(h, 0, 0, hours)
            self.get_miniute(num - h, 0, 0, minutes)
            for hour in hours:
                for miniute in minutes:
                    result.append(str(hour)+':'+str(miniute).zfill(2))
        return result
        
    def get_hour(self,  n, idx, tmp, result):
        if n == 0:
            result.append(tmp)
            return 
        for i in xrange(idx, 4):
            if tmp ^ (1<<i) > 11:
                return 
            else:
                self.get_hour(n-1, i+1, tmp^(1<<i), result)
                    
            
    def get_miniute(self, n, idx, tmp, result):
        if n == 0:
            result.append(tmp)
            return 
        for i in xrange(idx, 6):
            if tmp ^ (1<<i) > 59:
                return 
            else:
                self.get_miniute(n-1, i+1, tmp^(1<<i), result)
            
    

File: Algorithm/Java/401. Binary Watch.java
/*
* @Author: WuLC
* @Date:   2017-07-31 13:07:54
* @Last Modified by:   WuLC
* @Last Modified time: 2017-07-31 13:08:37
* @Email: liangchaowu5@gmail.com
*/

// dfs
public class Solution 
{
    private List<Integer> hours;
    private List<Integer> miniutes;
    public List<String> readBinaryWatch(int num) 
    {
        List<String> result = new ArrayList<String>();
        for (int i = 0; i < 4; i++)
        {
            hours = new ArrayList<Integer>();
            miniutes = new ArrayList<Integer>();
            getHour(i, 0, 0);
            getMiniute(num - i, 0, 0);
            String miniute;
            int h, m;
            for(int j = 0; j < hours.size(); j++)
            {
                for(int k = 0; k < miniutes.size(); k++)
                {
                    h = hours.get(j);
                    m = miniutes.get(k);
                    if(m < 10) result.add(String.valueOf(h)+":0"+String.valueOf(m));
                    else result.add(String.valueOf(h)+":"+String.valueOf(m));
                }    
            }
        }
        return result;
    }
    
    public void getHour(int n, int idx, int val)
    {
        if (n == 0 && val <= 11)
        {
            hours.add(val);
            return;
        }
        for(int i = idx; i < 4 ;i++)
            if ((val ^ (1 << idx)) <= 11) 
                getHour(n - 1, i + 1, val ^ (1 << i));
    }
    
    public void getMiniute(int n, int idx, int val)
    {
        if (n == 0 && val <= 59)
        {
            miniutes.add(val);
            return;
        }
        for(int i = idx; i < 6 ;i++)
            if ((val ^ (1 << idx)) <= 59) 
                getMiniute(n - 1, i + 1, val ^ (1 << i));
        
    }
}